ext/sodium/libsodium.c

```diff
@@ -0,0 +1,2666 @@
+						   crypto_stream_NONCEBYTES, CONST_CS | CONST_PERSISTENT);
+	REGISTER_LONG_CONSTANT("SODIUM_CRYPTO_STREAM_KEYBYTES",
+						   crypto_stream_KEYBYTES, CONST_CS | CONST_PERSISTENT);
+	return SUCCESS;
+}
+
+PHP_MSHUTDOWN_FUNCTION(sodium)
+{
+	return SUCCESS;
+}
+
+PHP_MINFO_FUNCTION(sodium)
+{
+	php_info_print_table_start();
+	php_info_print_table_header(2, "sodium support", "enabled");
+	php_info_print_table_header(2, "libsodium headers version", SODIUM_VERSION_STRING);
+	php_info_print_table_header(2, "libsodium library version", sodium_version_string());
+	php_info_print_table_end();
+}
+
+PHP_FUNCTION(sodium_memzero)
+{
+	zval	  *buf_zv;
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(),
+							  "z", &buf_zv) == FAILURE) {
+		return;
+	}
+	ZVAL_DEREF(buf_zv);
+	if (Z_TYPE_P(buf_zv) != IS_STRING) {
+		zend_throw_exception(sodium_exception_ce, "memzero: a PHP string is required", 0);
+		return;
+	}
+	if (Z_REFCOUNTED_P(buf_zv) && Z_REFCOUNT_P(buf_zv) == 1) {
+		char *buf = Z_STRVAL(*buf_zv);
+		size_t buf_len = Z_STRLEN(*buf_zv);
+		if (buf_len > 0) {
+			sodium_memzero(buf, (size_t) buf_len);
+		}
+	}
+	convert_to_null(buf_zv);
+}
+
+PHP_FUNCTION(sodium_increment)
+{
+	zval		  *val_zv;
+	unsigned char *val;
+	size_t		   i;
+	size_t		   val_len;
+	unsigned int   c;
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(),
+							  "z", &val_zv) == FAILURE) {
+		return;
+	}
+	ZVAL_DEREF(val_zv);
+	if (Z_TYPE_P(val_zv) != IS_STRING) {
+		zend_throw_exception(sodium_exception_ce, "increment(): a PHP string is required", 0);
+		return;
+	}
+
+	sodium_separate_string(val_zv);
+	val = (unsigned char *) Z_STRVAL(*val_zv);
+	val_len = Z_STRLEN(*val_zv);
+	c = 1U << 8;
+	for (i = (size_t) 0U; i < val_len; i++) {
+		c >>= 8;
+		c += val[i];
+		val[i] = (unsigned char) c;
+	}
+}
+
+PHP_FUNCTION(sodium_add)
+{
+	zval		  *val_zv;
+	unsigned char *val;
+	unsigned char *addv;
+	size_t		   i;
+	size_t		   val_len;
+	size_t		   addv_len;
+	unsigned int   c;
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(),
+							  "zs", &val_zv, &addv, &addv_len) == FAILURE) {
+		return;
+	}
+	ZVAL_DEREF(val_zv);
+	if (Z_TYPE_P(val_zv) != IS_STRING) {
+		zend_throw_exception(sodium_exception_ce, "add(): PHP strings are required", 0);
+		return;
+	}
+
+	sodium_separate_string(val_zv);
+	val = (unsigned char *) Z_STRVAL(*val_zv);
+	val_len = Z_STRLEN(*val_zv);
+	if (val_len != addv_len) {
+		zend_throw_exception(sodium_exception_ce, "add(): values must have the same length", 0);
+		return;
+	}
+	c = 0U;
+	for (i = (size_t) 0U; i < val_len; i++) {
+		c += val[i] + addv[i];
+		val[i] = (unsigned char) c;
+		c >>= 8;
+	}
+}
+
+PHP_FUNCTION(sodium_memcmp)
+{
+	char	  *buf1;
+	char	  *buf2;
+	size_t	   len1;
+	size_t	   len2;
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "ss",
+							  &buf1, &len1,
+							  &buf2, &len2) == FAILURE) {
+		return;
+	}
+	if (len1 != len2) {
+		zend_throw_exception(sodium_exception_ce, "memcmp(): arguments have different sizes", 0);
```

Comment:
I think the message should be `"memcmp(): arguments should have same size"`, since we tell users what we except, like other error messages in this file.